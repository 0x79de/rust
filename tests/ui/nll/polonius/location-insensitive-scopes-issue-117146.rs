// This is a non-regression test for issue #117146, where NLL and `-Zpolonius=next` computed
// different loan scopes when a region flowed into an SCC whose representative was an existential
// region.

//@ revisions: nll polonius
//@ [polonius] compile-flags: -Zpolonius=next

fn main() {
    let a = ();
    let b = |_| &a;
    //[nll]~^ ERROR closure may outlive the current function, but it borrows `a`
    //[polonius]~^^ ERROR closure may outlive the current function, but it borrows `a`
    bad(&b);
    //[nll]~^ ERROR implementation of `Fn`
    //[nll]~| ERROR implementation of `FnOnce`
    //[polonius]~^^^ ERROR implementation of `Fn`
    //[polonius]~| ERROR implementation of `FnOnce`
}

fn bad<F: Fn(&()) -> &()>(_: F) {}
