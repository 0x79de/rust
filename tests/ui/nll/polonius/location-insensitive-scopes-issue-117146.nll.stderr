error[E0373]: closure may outlive the current function, but it borrows `a`, which is owned by the current function
  --> $DIR/location-insensitive-scopes-issue-117146.rs:10:13
   |
LL |     let b = |_| &a;
   |             ^^^  - `a` is borrowed here
   |             |
   |             may outlive borrowed value `a`
   |
note: function requires argument type to outlive `'static`
  --> $DIR/location-insensitive-scopes-issue-117146.rs:13:5
   |
LL |     bad(&b);
   |     ^^^^^^^
help: to force the closure to take ownership of `a` (and any other referenced variables), use the `move` keyword
   |
LL |     let b = move |_| &a;
   |             ++++

error: implementation of `Fn` is not general enough
  --> $DIR/location-insensitive-scopes-issue-117146.rs:13:5
   |
LL |     bad(&b);
   |     ^^^^^^^ implementation of `Fn` is not general enough
   |
   = note: closure with signature `fn(&'2 ()) -> &()` must implement `Fn<(&'1 (),)>`, for any lifetime `'1`...
   = note: ...but it actually implements `Fn<(&'2 (),)>`, for some specific lifetime `'2`

error: implementation of `FnOnce` is not general enough
  --> $DIR/location-insensitive-scopes-issue-117146.rs:13:5
   |
LL |     bad(&b);
   |     ^^^^^^^ implementation of `FnOnce` is not general enough
   |
   = note: closure with signature `fn(&'2 ()) -> &()` must implement `FnOnce<(&'1 (),)>`, for any lifetime `'1`...
   = note: ...but it actually implements `FnOnce<(&'2 (),)>`, for some specific lifetime `'2`

error: aborting due to 3 previous errors

For more information about this error, try `rustc --explain E0373`.
